File: Data Structures & Algorithms/binary-tree-diameter/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int height(TreeNode* root){
        if(!root){
            return 0;
        }
        return max(height(root->left),height(root->right))+1;
    }

    int dfs(TreeNode* root,int &res){
        if(!root){
            return 0;
        }
        int l = dfs(root->left,res);
        int r = dfs(root->right,res);
        res = max(res,l+r);
        return max(l,r)+1;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        if(!root){
            return 0;
        }

        // int diameter = height(root->left)+height(root->right);
        // return max(diameter,max(
        //     diameterOfBinaryTree(root->left),
        //     diameterOfBinaryTree(root->right)
        //     )
        //     );
        int res=0;
        dfs(root,res);
        return res;
    }
};

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isLeaf(self,root)->bool:
        return not root.left and not root.right
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        if key<root.val:
           root.left = self.deleteNode(root.left,key)
        elif key>root.val:
           root.right = self.deleteNode(root.right,key)
        else:
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            
            prev = root
            successor = root.right
            while successor.left:
                prev = successor
                successor = successor.left
            root.val = successor.val
            if successor == prev.left:
                prev.left = self.deleteNode(successor,successor.val)
            else:
                prev.right = self.deleteNode(successor,successor.val)
                
        return root